qa/python_models/dlpack_test/model.py

```diff
@@ -99,28 +119,172 @@ def test_dlpack_gpu_tensors(self):
+    
+    def test_bool_datatype(self):
+        # [FIXME] pass bool_array directly to `pb_utils.Tensor.from_dlpack`,
+        # when numpy release supports DLPack bool type
+        bool_array = np.asarray([False, True])
+        bool_tensor = pb_utils.Tensor('tensor', bool_array)
+        bool_tensor_dlpack = pb_utils.Tensor.from_dlpack('tensor', bool_tensor)
+        self.assertTrue(
+            np.array_equal(bool_array, bool_tensor_dlpack.as_numpy()))
+    
+    def test_cuda_multi_stream(self):
+        # Test that external stream syncs with the default
+        # and pb_tensor has proper data
+        s1 = torch.cuda.Stream()
+        size = 5000
+        pytorch_tensor = torch.tensor([0,0,0,0], device='cuda')
+        expected_output = torch.tensor([2,2,2,2], device='cuda')
+        with torch.cuda.stream(s1):
+            matrix_a = torch.randn(size,size, device='cuda')
+            res = torch.matmul(matrix_a,matrix_a)
+            for _ in range(1000):
+                res = torch.matmul(res,matrix_a)
+            pytorch_tensor += torch.tensor([2,2,2,2], device='cuda')
+        
+        pb_tensor = pb_utils.Tensor.from_dlpack('tensor', pytorch_tensor)
+        pytorch_tensor_dlpack = from_dlpack(pb_tensor)
+        self.assertTrue(torch.equal(pytorch_tensor_dlpack, expected_output))
+    
+    def test_cuda_non_blocking_multi_stream(self):
+        # Test that external non-blocking stream syncs with the default stream
+        # and pb_tensor has proper data
+        non_blocking_stream = cp.cuda.Stream(non_blocking=True)
+        size = 5000
+        cupy_tensor = cp.array([0,0,0,0])
+        expected_output = cp.array([2,2,2,2])
+        with non_blocking_stream:
+            matrix_a = cp.random.rand(size,size)
+            res = cp.matmul(matrix_a,matrix_a)
+            for _ in range(1000):
+                res = cp.matmul(res,matrix_a)
+            cupy_tensor += cp.array([2,2,2,2])
+        
+        pb_tensor = pb_utils.Tensor.from_dlpack('tensor', cupy_tensor)
+        # Verify that non-blocking stream has no pending jobs left
+        self.assertTrue(non_blocking_stream.done)
+        cupy_tensor_dlpack = cp.from_dlpack(pb_tensor)
+        self.assertTrue(cp.array_equal(cupy_tensor_dlpack, expected_output))
+        self.assertFalse(pb_tensor.is_cpu())
+        self.assertEqual(
+            pb_tensor.__dlpack_device__(), cupy_tensor.__dlpack_device__())
+        
+    def test_cuda_multi_gpu(self):
+        # Test that pb_tensor on different device
+        size = 5000
+        expected_output = cp.array([2,2,2,2])
+        expected_dlpack_device = (2, 1) #  DLDeviceType::kDLCUDA, device_id 1
+        with cp.cuda.Device(1):
+            cupy_tensor = cp.array([0,0,0,0])
+            matrix_a = cp.random.rand(size,size)
+            res = cp.matmul(matrix_a,matrix_a)
+            for _ in range(1000):
+                res = cp.matmul(res,matrix_a)
```

Comment:
Does Cupy block on this line or it schedules the work asynchrounsly?

Reply:
I am not sure what you mean by `block`. I believe the scheduling is always sequential, since it happens on CPU.

This test schedules work on 2 different GPUs, since I don't create separate streams, the work is done on the default stream. I am checking that, when I am trying to consume capsule on GPU 0, ` pb_utils.Tensor.from_dlpack` switches to device 1 and waits for the work to finish on the device (see [here](https://github.com/triton-inference-server/python_backend/pull/223/files#diff-a6a4001d0fb912c1e37fd5d1ebac688a8e793c6b4f657d0e3c6d7de03282daffR371)). Basically the only way to do that (as I see it) is through the Device Sync. In this case, we don't care how the work is scheduled and at what time Device 0 requested the capsule, we still wait for the work to finish. This is a very strict way to do that, which I discussed [here](https://github.com/triton-inference-server/python_backend/pull/223#issuecomment-1531839008). It may happen that we will wait for the work to finish, when it is unnecessary. I've suggested one more addition to the code (a flag to make device sync controllable by the user), I would like to discuss that before though.

Reply:
I meant for example if you call `cp.matmul` is it going to schedule the `matmul` on a separate stream and immediately return or wait until the `matmul` is complete.

Reply:
I meant for example if you call `cp.matmul` is it going to schedule the `matmul` on a separate stream and immediately return or wait until the `matmul` is complete.